design/metal3host-controller.md

```diff
@@ -0,0 +1,314 @@
+# Baremetal host single instance and multi-instance operator
+
+## Introduction
+
+Currently, the metal3 system is to set up a Kubernetes cluster on bare-metal hosts which is tightly bound with cluster-api and Kubernetes. We target to allocate bare-metal hosts that just pre-installing the OS, or pre-installing the applications as required.
+
+## Motivation
+
+To meet the scenery that only requires a pure bare-metal host or with a simple application installed.
+
+## Goals
+
+- Allocate one pure bare-metal host
+- Allocate multiple bare-metal hosts
+- Allocate bare-metal host with pre-installed applications
+- Specify the different configurations for different bare-metal host
+
+## Non-goals
+
+- Allocate different types of bare-metal hosts at one time
+
+## Proposal
+
+### Architecture
+
+The operators include the metal3Host operator, metal3HostDeployment operator, and data operator.
+![modules](images/metal3hostModules.png)
+
+The bare-metal host controller and the ironic are not newly developed.
+
+Metal3Host is a virtual instance. Users could define it to require a bare-metal host. In the definition, the user could specify the image to install, could set the selector to choose a specific bare-metal host, and could define the boot data to automatically start their own applications.
+
+Metal3HostDeployment will define a group of metal3Hosts with the same configuration. It could define the number of replicas. It could scale up/down. We even could auto-scale by some rules after the monitor function is leveraged.
+
+The data operator will set up the boot time configuration & routine. It includes three parts: the userData, the metaData, and the networkData. The userData includes some user-specified data such as installing a package, or running a command like "mkdir /test". The metaData defines some host-specific data such as hostname. The networkData includes the network configuration. The data is bound with the baremetal host. For example, the networkData defines a NIC with an IP, the NIC name is host related. So a dataTemplate should be pre-defined, it will be bound to the bare-metal host when a metal3host is defined and associated with a bare-metal host.
```

Comment:
It is not clear to me why do you need a new pair of data and dataTemplate CR, wouldn't it be easier to just re-use the existing metal3Data and metal3DataTemplates?

You would also use the new data and the dataTemplate for the same config process as the existing metal3Data and metal3DataTemplate is used for.

Reply:
Yes, almost the same with metal3Data. One reason I want to create a new CR is, the metal3Data is bound to the metal3Machine, so it's not easy to modify if any change is required. Another is that the metal3Host and data could be an independent application(new a repo).
Maybe I'm wrong about this.